File: CSharpRag/CSharpRagApp/Services/CitationParser.cs
using System.Text.Json;

namespace CSharpRagApp.Services;

/// <summary>
/// Service for parsing citations from RAG responses
/// </summary>
public interface ICitationParser
{
    List<Citation> ParseCitations(string jsonResponse);
    string FormatCitation(Citation citation, int number);
    string TruncateContent(string content, int maxLength = 150);
}

/// <summary>
/// Represents a citation from a RAG response
/// </summary>
public class Citation
{
    public string? Title { get; set; }
    public string? FilePath { get; set; }
    public string? Content { get; set; }
    
    public string DisplayTitle => Title ?? FilePath ?? "Unknown Document";
}

public class CitationParser : ICitationParser
{
    public List<Citation> ParseCitations(string jsonResponse)
    {
        var citations = new List<Citation>();
        
        if (string.IsNullOrEmpty(jsonResponse))
            return citations;

        try
        {
            using var doc = JsonDocument.Parse(jsonResponse);
            
            if (!doc.RootElement.TryGetProperty("choices", out var choices) || 
                choices.GetArrayLength() == 0)
            {
                return citations;
            }

            var firstChoice = choices[0];
            if (!firstChoice.TryGetProperty("message", out var message) ||
                !message.TryGetProperty("context", out var context) ||
                !context.TryGetProperty("citations", out var citationsArray))
            {
                return citations;
            }

            foreach (var citationElement in citationsArray.EnumerateArray())
            {
                var citation = new Citation
                {
                    Title = citationElement.TryGetProperty("title", out var t) ? t.GetString() : null,
                    FilePath = citationElement.TryGetProperty("filepath", out var f) ? f.GetString() : null,
                    Content = citationElement.TryGetProperty("content", out var c) ? c.GetString() : null
                };
                citations.Add(citation);
            }
        }
        catch (JsonException)
        {
            // Return empty list on parse errors
        }

        return citations;
    }

    public string FormatCitation(Citation citation, int number)
    {
        var displayTitle = citation.Title ?? citation.FilePath ?? $"Document {number}";
        var result = $"  [{number}] {displayTitle}";

        if (!string.IsNullOrEmpty(citation.Content))
        {
            var preview = TruncateContent(citation.Content);
            result += $"\n      {preview}";
        }

        return result;
    }

    public string TruncateContent(string content, int maxLength = 150)
    {
        if (string.IsNullOrEmpty(content))
            return string.Empty;

        if (content.Length <= maxLength)
            return content;

        return content.Substring(0, maxLength) + "...";
    }
}

File: PythonRag/python_rag_app/services/citation_parser.py
"""Citation parser service for RAG responses."""

import json
from dataclasses import dataclass
from typing import Protocol


@dataclass
class Citation:
    """Represents a citation from a RAG response."""

    title: str | None = None
    file_path: str | None = None
    content: str | None = None

    @property
    def display_title(self) -> str:
        """Get display title for the citation."""
        return self.title or self.file_path or "Unknown Document"


class ICitationParser(Protocol):
    """Interface for citation parser."""

    def parse_citations(self, json_response: str) -> list[Citation]:
        """Parse citations from JSON response."""
        ...

    def format_citation(self, citation: Citation, number: int) -> str:
        """Format a citation for display."""
        ...

    def truncate_content(self, content: str, max_length: int = 150) -> str:
        """Truncate content to max length."""
        ...


class CitationParser:
    """Service for parsing citations from RAG responses."""

    def parse_citations(self, json_response: str) -> list[Citation]:
        """Parse citations from a JSON response.

        Args:
            json_response: JSON string from RAG response.

        Returns:
            List of Citation objects.
        """
        citations = []

        if not json_response:
            return citations

        try:
            data = json.loads(json_response)

            choices = data.get("choices", [])
            if not choices:
                return citations

            first_choice = choices[0]
            message = first_choice.get("message", {})
            context = message.get("context", {})
            citations_array = context.get("citations", [])

            for citation_element in citations_array:
                citation = Citation(
                    title=citation_element.get("title"),
                    file_path=citation_element.get("filepath"),
                    content=citation_element.get("content"),
                )
                citations.append(citation)

        except (json.JSONDecodeError, KeyError, IndexError):
            # Return empty list on parse errors
            pass

        return citations

    def format_citation(self, citation: Citation, number: int) -> str:
        """Format a citation for display.

        Args:
            citation: Citation to format.
            number: Citation number.

        Returns:
            Formatted citation string.
        """
        display_title = citation.title or citation.file_path or f"Document {number}"
        result = f"  [{number}] {display_title}"

        if citation.content:
            preview = self.truncate_content(citation.content)
            result += f"\n      {preview}"

        return result

    def truncate_content(self, content: str, max_length: int = 150) -> str:
        """Truncate content to a maximum length.

        Args:
            content: Content to truncate.
            max_length: Maximum length (default: 150).

        Returns:
            Truncated content with ellipsis if needed.
        """
        if not content:
            return ""

        if len(content) <= max_length:
            return content

        return content[:max_length] + "..."
